docs/rholang/rholangmatchingtut.md

```diff
@@ -0,0 +1,181 @@
+
+Furthermore, we can't bind variables to parts of patterns. For example, the following send/receive will not match:
+
+    1 for( @ for( @{x!(y)} <- @Nil ){ Nil } <- @Nil ){ Nil }  |
+    2 @Nil!( for( @{x!(10)} <- @Nil ){ Nil } )
+
+Na&iuml;vely, one might expect these to match, binding `y` to `10`, but to match with the above receive, one must send something alpha equivalent to:
+
+    @Nil!( for( @x!(y) <- @Nil ){ Nil } )
+
+## Precedence Rules
+When writing more complicated patterns, one should be aware of precedence rules of the operators in Rholang. Take for example the following pattern:
+
+    @{@x!(Nil) | y!(Nil)}
+
+A priori, it is unclear whether this should be interpreted as `@{{@x!(Nil)} | {y!(Nil)}}`, where `x` is a process variable and `y` is a name variable, or as `@{@{x!(Nil) | y}!(Nil)}}`, where `x` is a name variable and `y` is a process variable. The chosen interpretation is vital, since the way we use `x` and `y` in the body depend on the type (process or name) of each of these terms. If the first interpretation is correct,
+
+    for( @{@x!(Nil) | y!(Nil)} <- @10 ){ @x!("success") | y!("success") }
+
+will compile, while
+
+    for( @{@x!(Nil) | y!(Nil)} <- @10 ){ x!("success") | @y!("success") }
+
+will not. If the second interpretation is correct, the opposite is true.
+
+In Rholang, `@` binds tighter than `|`. Thus our first interpretation, `@{{@x!(Nil)} | {y!(Nil)}}`, is correct. We also have precedence rules for logical operations. Take for example the pattern
+
+    x /\ @y!(Nil) \/ @z!(10)
+
+Like in the example above, we could either interpret this as `{x /\ @y!(Nil)} \/ @z!(10)`, or as `x /\ {@y!(Nil) \/ @z!(10)}`. In Rholang, `/\` takes precedence over `\/` (as is the standard for logical connectives).
+
+## Illegal Moves
+There are some illegal moves that we ought to cover. The first has to do with arithmetic operations which, if you recall from the section on name equivalence, are evaluated on the top level, but not anywhere else. Because of this, in Rholang we cannot match parts of arithmetic operations. For example, we might expect
+
+    for( @{x + 7} <- @Nil ){ Nil }
+
+to match with a send such as `@Nil!(5 + 7)`, binding `x` to `5`. However, when messages are sent, top-level expressions are evaluated before sending, so this send will not match with anything&mdash;although it will compile!
+
+Remember, however, that an arithmetic operation that is in a pattern within a pattern is not evaluated when sent over a channel, and must be matched exactly. For example,
+
+    @Nil!( for( x <- @{5 + 7}){Nil} )
+
+can only match with something that preserves the `5 + 7` intact, such as in the process `for( @for( x <- @{5 + 7}){z} <- @Nil ){ @z!(Nil) }`. Since the `5 + 7` is part of a pattern within a pattern, we still cannot match any part of it.
```

Comment:
Is the right side of a `for` statement considered a pattern? So IIUC, 

for (pattern <- pattern) { not-a-pattern }

?

Reply:
No, these are of the form `for( pattern <- name ) { process }`.

Reply:
No, these are of the form `for( pattern <- name ){ process }`. This was worded really poorly, and inconsistent with the rest of the document. Just committed a change--let me know if what I wrote is clearer now.